[InstCombine] Canonicalize more saturated-add variants

AZero13 opened the pull request:
LLVM is not evaluating X u > C, a, b the same way it evaluates X <= C, b, a.

To fix this, let's move the folds to after the canonicalization of -1 to TrueVal.

Let's allow splat vectors with poison elements to be recognized too!

As a result, negative tests that didn't get folded due to having poison in the vector now fold as well!

Finally, for completion, handle the one case that isn't caught by the above checks because it is canonicalized to eq:
X == -1 ? -1 : X + 1 -> uadd.sat(X, 1)

Alive2 Proof:
https://alive2.llvm.org/ce/z/WEcgYH

AZero13 committed:
Pre-commit tests (NFC)

goldsteinn commented:
Why did you close https://github.com/llvm/llvm-project/pull/97973?

AtariDreams commented:
> Why did you close #97973?

Wouldn't let me reopen after a botched rebase.

nikic reviewed (commented):
Please replace tests using something like uge C with ugt C-1 so the input is canonical. And then remove all the tests that become duplicates as a result.

Your alive2 proofs also include some preconditions -- some cases check that C!=0. But it does not look like your implementation contains those checks?

AtariDreams commented:
> Please replace tests using something like uge C with ugt C-1 so the input is canonical. And then remove all the tests that become duplicates as a result.
> 
> 
> 
> Your alive2 proofs also include some preconditions -- some cases check that C!=0. But it does not look like your implementation contains those checks?

0 gets canonicalized out anyway so it won't be caught. That was just for the sake of alive2

AtariDreams commented:
> Please replace tests using something like uge C with ugt C-1 so the input is canonical. And then remove all the tests that become duplicates as a result.
> 
> Your alive2 proofs also include some preconditions -- some cases check that C!=0. But it does not look like your implementation contains those checks?

Fixed!

AtariDreams commented:
@nikic I addressed the issues

AtariDreams commented:
> Please replace tests using something like uge C with ugt C-1 so the input is canonical. And then remove all the tests that become duplicates as a result.
> 
> Your alive2 proofs also include some preconditions -- some cases check that C!=0. But it does not look like your implementation contains those checks?

It does now!

AtariDreams commented:
@topperc @goldsteinn Do any of you have objections to this?

topperc commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -996,6 +989,30 @@ static Value *canonicalizeSaturatedAdd(ICmpInst *Cmp, Value *TVal, Value *FVal,
 
+  if ((Pred == ICmpInst::ICMP_UGE || Pred == ICmpInst::ICMP_UGT) &&
+      match(FVal, m_Add(m_Specific(Cmp0), m_APInt(C))) &&
+      match(Cmp1, m_SpecificIntAllowPoison(~*C))) {
```
m_SpecificIntAllowPoison allows poison elements in splat vectors. It doesn't do anything for scalars. All of your tests appear to be scalar. So the use of `AllowPoison` isn't tested?

RSilicon replied:
I did test AllowPoison and the end result doesn't change.

RSilicon replied:
Perhaps this doesn't do anything for poison since it already folded to poison. Are poison in splat vectors just vectors of nothing but poison?

RSilicon replied:
See: uadd_sat_poison

topperc replied:
uadd_sat_poison is a scalar poison value. m_SpecifincIntAllowPoison won't match that.

m_SpecificIntAllowPoison allows matching a vector where some elements are poison and some are the specific constant like `<i32 31, i32 poison, i32 31, i32 31>`. For scalars it only matches the specific constant.

RSilicon replied:
Allow me to update the tests then.

RSilicon replied:
I updated the tests to reflect this!

RSilicon replied:
> uadd_sat_poison is a scalar poison value. m_SpecifincIntAllowPoison won't match that.
> 
> m_SpecificIntAllowPoison allows matching a vector where some elements are poison and some are the specific constant like `<i32 31, i32 poison, i32 31, i32 31>`. For scalars it only matches the specific constant.

I added a splat-poison vector test to show this!

nikic commented:
This PR was pretty close to an acceptable state the last time I looked at it. Now you have once again added extra changes that nobody asked for, and moved it away from acceptance again.

RSilicon commented:
> This PR was pretty close to an acceptable state the last time I looked at it. Now you have once again added extra changes that nobody asked for, and moved it away from acceptance again.

The change to AllowPoison is warranted. See the tests and Alive2.

RSilicon commented:
> This PR was pretty close to an acceptable state the last time I looked at it. Now you have once again added extra changes that nobody asked for, and moved it away from acceptance again.

And with that change, I fully exhausted all possible canonicalizations, well, minus actual variables but that will be for a follow-up PR

RSilicon commented:
> This PR was pretty close to an acceptable state the last time I looked at it. Now you have once again added extra changes that nobody asked for, and moved it away from acceptance again.

Well, it is now exhaustive, so the good news is that this is as far as it is gonna get.

RSilicon commented:
@topperc What do you think about this?

topperc commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -996,6 +989,30 @@ static Value *canonicalizeSaturatedAdd(ICmpInst *Cmp, Value *TVal, Value *FVal,
+  if ((Pred == ICmpInst::ICMP_UGE || Pred == ICmpInst::ICMP_UGT) &&
+      match(FVal, m_Add(m_Specific(Cmp0), m_APIntAllowPoison(C))) &&
+      match(Cmp1, m_SpecificIntAllowPoison(~*C))) {
+    // (X u> ~C) ? -1 : (X + C) --> uadd.sat(X, C)
```
need a comment for `(X u>= ~C) ? -1 : (X + C) --> uadd.sat(X, C)` too

RSilicon replied:
> need a comment for `(X u>= ~C) ? -1 : (X + C) --> uadd.sat(X, C)` too

Done!

topperc commented:
Your alive proof has

```
define dso_local i8 @src3(i8 %x, i8  %c) {
entry:
  %noZero = icmp ne i8 %c, 0
  call void @llvm.assume(i1 %noZero)
  %add = add nuw i8 %x, %c
  %c.not = xor i8 %c, -1
  %d = add i8 %c.not, 1
  %cmp.not = icmp ugt i8 %x, %d
  %cond = select i1 %cmp.not, i8 -1, i8 %add
  ret i8 %cond
}
```

The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.

topperc commented:
Your proof has this

```
define dso_local i8 @src2(i8 %x, i8  %c) {
entry:
  %add = add i8 %x, %c
  %c.not = xor i8 %c, -1
  %d = sub nuw i8 %c.not, 1
  %cmp.not = icmp ugt i8 %x, %d
  %cond = select i1 %cmp.not, i8 -1, i8 %add
  ret i8 %cond
}
```

Which has an nuw on the sub. It fails without it. Your transform does not check this condition.

RSilicon commented:
> Your alive proof has
> 
> ```
> define dso_local i8 @src3(i8 %x, i8  %c) {
> entry:
>   %noZero = icmp ne i8 %c, 0
>   call void @llvm.assume(i1 %noZero)
>   %add = add nuw i8 %x, %c
>   %c.not = xor i8 %c, -1
>   %d = add i8 %c.not, 1
>   %cmp.not = icmp ugt i8 %x, %d
>   %cond = select i1 %cmp.not, i8 -1, i8 %add
>   ret i8 %cond
> }
> ```
> 
> The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.

I meant to remove this. This exact situation is impossible to run into in real code. It is not covered by my code at all.

RSilicon commented:
> Your proof has this
> 
> ```
> define dso_local i8 @src2(i8 %x, i8  %c) {
> entry:
>   %add = add i8 %x, %c
>   %c.not = xor i8 %c, -1
>   %d = sub nuw i8 %c.not, 1
>   %cmp.not = icmp ugt i8 %x, %d
>   %cond = select i1 %cmp.not, i8 -1, i8 %add
>   ret i8 %cond
> }
> ```
> 
> Which has an `nuw` on the `sub`. It fails without it. Your transform does not check this condition.

This is also impossible to encounter. I was just testing all possible theoretical combinations.

RSilicon commented:
> Your proof has this
> 
> 
> 
> ```
> 
> define dso_local i8 @src2(i8 %x, i8  %c) {
> 
> entry:
> 
>   %add = add i8 %x, %c
> 
>   %c.not = xor i8 %c, -1
> 
>   %d = sub nuw i8 %c.not, 1
> 
>   %cmp.not = icmp ugt i8 %x, %d
> 
>   %cond = select i1 %cmp.not, i8 -1, i8 %add
> 
>   ret i8 %cond
> 
> }
> 
> ```
> 
> 
> 
> Which has an `nuw` on the `sub`. It fails without it. Your transform does not check this condition.

Because we are checking constants, and not variables, if this wraps, then m_SpecificInt will always fail.

RSilicon commented:
> Your proof has this
> 
> 
> 
> ```
> 
> define dso_local i8 @src2(i8 %x, i8  %c) {
> 
> entry:
> 
>   %add = add i8 %x, %c
> 
>   %c.not = xor i8 %c, -1
> 
>   %d = sub nuw i8 %c.not, 1
> 
>   %cmp.not = icmp ugt i8 %x, %d
> 
>   %cond = select i1 %cmp.not, i8 -1, i8 %add
> 
>   ret i8 %cond
> 
> }
> 
> ```
> 
> 
> 
> Which has an `nuw` on the `sub`. It fails without it. Your transform does not check this condition.

Ditto.

RSilicon commented:
> Your alive proof has
> 
> 
> 
> ```
> 
> define dso_local i8 @src3(i8 %x, i8  %c) {
> 
> entry:
> 
>   %noZero = icmp ne i8 %c, 0
> 
>   call void @llvm.assume(i1 %noZero)
> 
>   %add = add nuw i8 %x, %c
> 
>   %c.not = xor i8 %c, -1
> 
>   %d = add i8 %c.not, 1
> 
>   %cmp.not = icmp ugt i8 %x, %d
> 
>   %cond = select i1 %cmp.not, i8 -1, i8 %add
> 
>   ret i8 %cond
> 
> }
> 
> ```
> 
> 
> 
> The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.

Ditto.

RSilicon commented:
> Your alive proof has
> 
> 
> 
> ```
> 
> define dso_local i8 @src3(i8 %x, i8  %c) {
> 
> entry:
> 
>   %noZero = icmp ne i8 %c, 0
> 
>   call void @llvm.assume(i1 %noZero)
> 
>   %add = add nuw i8 %x, %c
> 
>   %c.not = xor i8 %c, -1
> 
>   %d = add i8 %c.not, 1
> 
>   %cmp.not = icmp ugt i8 %x, %d
> 
>   %cond = select i1 %cmp.not, i8 -1, i8 %add
> 
>   ret i8 %cond
> 
> }
> 
> ```
> 
> 
> 
> The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.

x ugt -C ? x + C: -1 always saturates, but nuw means it doesn't, so this would be malformed IR.

RSilicon commented:
> Your alive proof has
> 
> 
> 
> ```
> 
> define dso_local i8 @src3(i8 %x, i8  %c) {
> 
> entry:
> 
>   %noZero = icmp ne i8 %c, 0
> 
>   call void @llvm.assume(i1 %noZero)
> 
>   %add = add nuw i8 %x, %c
> 
>   %c.not = xor i8 %c, -1
> 
>   %d = add i8 %c.not, 1
> 
>   %cmp.not = icmp ugt i8 %x, %d
> 
>   %cond = select i1 %cmp.not, i8 -1, i8 %add
> 
>   ret i8 %cond
> 
> }
> 
> ```
> 
> 
> 
> The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.

This shouldn't be in my proof. I apologize: it's actually malformed IR.

nuw is impossible for x ugt -C ? x + C : -1;

RSilicon commented:
I removed the nuw cases from my proof because they are actually malformed IR and are not covered at all as they are impossible to run into assuming the IR isn't malformed.

topperc commented:
> > Your alive proof has
> > 
> > 
> > 
> > ```
> > 
> > define dso_local i8 @src3(i8 %x, i8  %c) {
> > 
> > entry:
> > 
> >   %noZero = icmp ne i8 %c, 0
> > 
> >   call void @llvm.assume(i1 %noZero)
> > 
> >   %add = add nuw i8 %x, %c
> > 
> >   %c.not = xor i8 %c, -1
> > 
> >   %d = add i8 %c.not, 1
> > 
> >   %cmp.not = icmp ugt i8 %x, %d
> > 
> >   %cond = select i1 %cmp.not, i8 -1, i8 %add
> > 
> >   ret i8 %cond
> > 
> > }
> > 
> > ```
> > 
> > 
> > 
> > The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.
> 
> This shouldn't be in my proof. I apologize: it's actually malformed IR.
> 
> nuw is impossible for x ugt -C ? x + C : -1;

Why would it be malformed IR? The select would pick -1 in the case of overflow and the add would produce poison. The selects blocks the poison from propagating by picking -1. That's completely valid IR

RSilicon commented:
Point is that LLVM would NEVER assign nuw to that case

RSilicon commented:
> > > Your alive proof has
> > > 
> > > 
> > > 
> > > ```
> > > 
> > > define dso_local i8 @src3(i8 %x, i8  %c) {
> > > 
> > > entry:
> > > 
> > >   %noZero = icmp ne i8 %c, 0
> > > 
> > >   call void @llvm.assume(i1 %noZero)
> > > 
> > >   %add = add nuw i8 %x, %c
> > > 
> > >   %c.not = xor i8 %c, -1
> > > 
> > >   %d = add i8 %c.not, 1
> > > 
> > >   %cmp.not = icmp ugt i8 %x, %d
> > > 
> > >   %cond = select i1 %cmp.not, i8 -1, i8 %add
> > > 
> > >   ret i8 %cond
> > > 
> > > }
> > > 
> > > ```
> > > 
> > > 
> > > 
> > > The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.
> > 
> > This shouldn't be in my proof. I apologize: it's actually malformed IR.
> > 
> > nuw is impossible for x ugt -C ? x + C : -1;
> 
> Why would it be malformed IR? The select would pick -1 in the case of overflow and the add would produce poison. The selects blocks the poison from propagating by picking -1. That's completely valid IR

LLVM would NEVER put nuw for x + C if x ugt -C was true. 

Let C1 be the positive interpretation of -C in two's compliment.

x ugt C being true means x is at (unsigned least) C1 + 1. But, C + C1 would wrap. meaning X + C MUST wrap if X ugt -C.

topperc commented:
> > > > Your alive proof has
> > > > ```
> > > > 
> > > > define dso_local i8 @src3(i8 %x, i8  %c) {
> > > > 
> > > > entry:
> > > > 
> > > >   %noZero = icmp ne i8 %c, 0
> > > > 
> > > >   call void @llvm.assume(i1 %noZero)
> > > > 
> > > >   %add = add nuw i8 %x, %c
> > > > 
> > > >   %c.not = xor i8 %c, -1
> > > > 
> > > >   %d = add i8 %c.not, 1
> > > > 
> > > >   %cmp.not = icmp ugt i8 %x, %d
> > > > 
> > > >   %cond = select i1 %cmp.not, i8 -1, i8 %add
> > > > 
> > > >   ret i8 %cond
> > > > 
> > > > }
> > > > ```
> > > > 
> > > > 
> > > >     
> > > >       
> > > >     
> > > > 
> > > >       
> > > >     
> > > > 
> > > >     
> > > >   
> > > > The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.
> > > 
> > > 
> > > This shouldn't be in my proof. I apologize: it's actually malformed IR.
> > > nuw is impossible for x ugt -C ? x + C : -1;
> > 
> > 
> > Why would it be malformed IR? The select would pick -1 in the case of overflow and the add would produce poison. The selects blocks the poison from propagating by picking -1. That's completely valid IR
> 
> LLVM would NEVER put nuw for x + C if x ugt -C was true.
> 
> Let C1 be the positive interpretation of -C in two's compliment.
> 
> x ugt C1 being true means x is at (unsigned least) C1 + 1. But, C + C1 would wrap. meaning X + C MUST wrap if X ugt -C.

Ignoring the undef input possibility, alive2 says its valid to apply nuw. https://alive2.llvm.org/ce/z/sPStwP

RSilicon commented:
> > > > > Your alive proof has
> > > > > ```
> > > > > 
> > > > > define dso_local i8 @src3(i8 %x, i8  %c) {
> > > > > 
> > > > > entry:
> > > > > 
> > > > >   %noZero = icmp ne i8 %c, 0
> > > > > 
> > > > >   call void @llvm.assume(i1 %noZero)
> > > > > 
> > > > >   %add = add nuw i8 %x, %c
> > > > > 
> > > > >   %c.not = xor i8 %c, -1
> > > > > 
> > > > >   %d = add i8 %c.not, 1
> > > > > 
> > > > >   %cmp.not = icmp ugt i8 %x, %d
> > > > > 
> > > > >   %cond = select i1 %cmp.not, i8 -1, i8 %add
> > > > > 
> > > > >   ret i8 %cond
> > > > > 
> > > > > }
> > > > > ```
> > > > > 
> > > > > 
> > > > >     
> > > > >       
> > > > >     
> > > > > 
> > > > >       
> > > > >     
> > > > > 
> > > > >     
> > > > >   
> > > > > The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.
> > > > 
> > > > 
> > > > This shouldn't be in my proof. I apologize: it's actually malformed IR.
> > > > nuw is impossible for x ugt -C ? x + C : -1;
> > > 
> > > 
> > > Why would it be malformed IR? The select would pick -1 in the case of overflow and the add would produce poison. The selects blocks the poison from propagating by picking -1. That's completely valid IR
> > 
> > 
> > LLVM would NEVER put nuw for x + C if x ugt -C was true.
> > Let C1 be the positive interpretation of -C in two's compliment.
> > x ugt C1 being true means x is at (unsigned least) C1 + 1. But, C + C1 would wrap. meaning X + C MUST wrap if X ugt -C.
> 
> Ignoring the undef input possibility, alive2 says its valid to apply nuw. https://alive2.llvm.org/ce/z/sPStwP

Okay thank you. Applying fixes right now.

topperc commented:
> > > > > > Your alive proof has
> > > > > > ```
> > > > > > 
> > > > > > define dso_local i8 @src3(i8 %x, i8  %c) {
> > > > > > 
> > > > > > entry:
> > > > > > 
> > > > > >   %noZero = icmp ne i8 %c, 0
> > > > > > 
> > > > > >   call void @llvm.assume(i1 %noZero)
> > > > > > 
> > > > > >   %add = add nuw i8 %x, %c
> > > > > > 
> > > > > >   %c.not = xor i8 %c, -1
> > > > > > 
> > > > > >   %d = add i8 %c.not, 1
> > > > > > 
> > > > > >   %cmp.not = icmp ugt i8 %x, %d
> > > > > > 
> > > > > >   %cond = select i1 %cmp.not, i8 -1, i8 %add
> > > > > > 
> > > > > >   ret i8 %cond
> > > > > > 
> > > > > > }
> > > > > > ```
> > > > > > 
> > > > > > 
> > > > > >     
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >       
> > > > > >     
> > > > > > 
> > > > > >     
> > > > > >   
> > > > > > The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.
> > > > > 
> > > > > 
> > > > > This shouldn't be in my proof. I apologize: it's actually malformed IR.
> > > > > nuw is impossible for x ugt -C ? x + C : -1;
> > > > 
> > > > 
> > > > Why would it be malformed IR? The select would pick -1 in the case of overflow and the add would produce poison. The selects blocks the poison from propagating by picking -1. That's completely valid IR
> > > 
> > > 
> > > LLVM would NEVER put nuw for x + C if x ugt -C was true.
> > > Let C1 be the positive interpretation of -C in two's compliment.
> > > x ugt C1 being true means x is at (unsigned least) C1 + 1. But, C + C1 would wrap. meaning X + C MUST wrap if X ugt -C.
> > 
> > 
> > Ignoring the undef input possibility, alive2 says its valid to apply nuw. https://alive2.llvm.org/ce/z/sPStwP
> 
> Okay thank you. Applying fixes right now.

What fix are you applying?

RSilicon commented:
> This PR was pretty close to an acceptable state the last time I looked at it. Now you have once again added extra changes that nobody asked for, and moved it away from acceptance again

> > > > > > > Your alive proof has
> > > > > > > ```
> > > > > > > 
> > > > > > > define dso_local i8 @src3(i8 %x, i8  %c) {
> > > > > > > 
> > > > > > > entry:
> > > > > > > 
> > > > > > >   %noZero = icmp ne i8 %c, 0
> > > > > > > 
> > > > > > >   call void @llvm.assume(i1 %noZero)
> > > > > > > 
> > > > > > >   %add = add nuw i8 %x, %c
> > > > > > > 
> > > > > > >   %c.not = xor i8 %c, -1
> > > > > > > 
> > > > > > >   %d = add i8 %c.not, 1
> > > > > > > 
> > > > > > >   %cmp.not = icmp ugt i8 %x, %d
> > > > > > > 
> > > > > > >   %cond = select i1 %cmp.not, i8 -1, i8 %add
> > > > > > > 
> > > > > > >   ret i8 %cond
> > > > > > > 
> > > > > > > }
> > > > > > > ```
> > > > > > > 
> > > > > > > 
> > > > > > >     
> > > > > > >       
> > > > > > >     
> > > > > > > 
> > > > > > >       
> > > > > > >     
> > > > > > > 
> > > > > > >     
> > > > > > >   
> > > > > > > The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.
> > > > > > 
> > > > > > 
> > > > > > This shouldn't be in my proof. I apologize: it's actually malformed IR.
> > > > > > nuw is impossible for x ugt -C ? x + C : -1;
> > > > > 
> > > > > 
> > > > > Why would it be malformed IR? The select would pick -1 in the case of overflow and the add would produce poison. The selects blocks the poison from propagating by picking -1. That's completely valid IR
> > > > 
> > > > 
> > > > LLVM would NEVER put nuw for x + C if x ugt -C was true.
> > > > Let C1 be the positive interpretation of -C in two's compliment.
> > > > x ugt C1 being true means x is at (unsigned least) C1 + 1. But, C + C1 would wrap. meaning X + C MUST wrap if X ugt -C.
> > > 
> > > 
> > > Ignoring the undef input possibility, alive2 says its valid to apply nuw. https://alive2.llvm.org/ce/z/sPStwP
> > 
> > 
> > Okay thank you. Applying fixes right now.
> 
> What fix are you applying?

I am going to apply the nuw stuff.

RSilicon commented:
> > This PR was pretty close to an acceptable state the last time I looked at it. Now you have once again added extra changes that nobody asked for, and moved it away from acceptance again
> 
> > > > > > > > Your alive proof has
> > > > > > > > ```
> > > > > > > > 
> > > > > > > > define dso_local i8 @src3(i8 %x, i8  %c) {
> > > > > > > > 
> > > > > > > > entry:
> > > > > > > > 
> > > > > > > >   %noZero = icmp ne i8 %c, 0
> > > > > > > > 
> > > > > > > >   call void @llvm.assume(i1 %noZero)
> > > > > > > > 
> > > > > > > >   %add = add nuw i8 %x, %c
> > > > > > > > 
> > > > > > > >   %c.not = xor i8 %c, -1
> > > > > > > > 
> > > > > > > >   %d = add i8 %c.not, 1
> > > > > > > > 
> > > > > > > >   %cmp.not = icmp ugt i8 %x, %d
> > > > > > > > 
> > > > > > > >   %cond = select i1 %cmp.not, i8 -1, i8 %add
> > > > > > > > 
> > > > > > > >   ret i8 %cond
> > > > > > > > 
> > > > > > > > }
> > > > > > > > ```
> > > > > > > > 
> > > > > > > > 
> > > > > > > >     
> > > > > > > >       
> > > > > > > >     
> > > > > > > > 
> > > > > > > >       
> > > > > > > >     
> > > > > > > > 
> > > > > > > >     
> > > > > > > >   
> > > > > > > > The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.
> > > > > > > 
> > > > > > > 
> > > > > > > This shouldn't be in my proof. I apologize: it's actually malformed IR.
> > > > > > > nuw is impossible for x ugt -C ? x + C : -1;
> > > > > > 
> > > > > > 
> > > > > > Why would it be malformed IR? The select would pick -1 in the case of overflow and the add would produce poison. The selects blocks the poison from propagating by picking -1. That's completely valid IR
> > > > > 
> > > > > 
> > > > > LLVM would NEVER put nuw for x + C if x ugt -C was true.
> > > > > Let C1 be the positive interpretation of -C in two's compliment.
> > > > > x ugt C1 being true means x is at (unsigned least) C1 + 1. But, C + C1 would wrap. meaning X + C MUST wrap if X ugt -C.
> > > > 
> > > > 
> > > > Ignoring the undef input possibility, alive2 says its valid to apply nuw. https://alive2.llvm.org/ce/z/sPStwP
> > > 
> > > 
> > > Okay thank you. Applying fixes right now.
> > 
> > 
> > What fix are you applying?
> 
> I am going to apply the nuw stuff. I just need to update my tests

Done!

topperc commented:
> > > This PR was pretty close to an acceptable state the last time I looked at it. Now you have once again added extra changes that nobody asked for, and moved it away from acceptance again
> 
> > 
> 
> > > > > > > > > Your alive proof has
> 
> > > > > > > > > ```
> 
> > > > > > > > > 
> 
> > > > > > > > > define dso_local i8 @src3(i8 %x, i8  %c) {
> 
> > > > > > > > > 
> 
> > > > > > > > > entry:
> 
> > > > > > > > > 
> 
> > > > > > > > >   %noZero = icmp ne i8 %c, 0
> 
> > > > > > > > > 
> 
> > > > > > > > >   call void @llvm.assume(i1 %noZero)
> 
> > > > > > > > > 
> 
> > > > > > > > >   %add = add nuw i8 %x, %c
> 
> > > > > > > > > 
> 
> > > > > > > > >   %c.not = xor i8 %c, -1
> 
> > > > > > > > > 
> 
> > > > > > > > >   %d = add i8 %c.not, 1
> 
> > > > > > > > > 
> 
> > > > > > > > >   %cmp.not = icmp ugt i8 %x, %d
> 
> > > > > > > > > 
> 
> > > > > > > > >   %cond = select i1 %cmp.not, i8 -1, i8 %add
> 
> > > > > > > > > 
> 
> > > > > > > > >   ret i8 %cond
> 
> > > > > > > > > 
> 
> > > > > > > > > }
> 
> > > > > > > > > ```
> 
> > > > > > > > > 
> 
> > > > > > > > > 
> 
> > > > > > > > >     
> 
> > > > > > > > >       
> 
> > > > > > > > >     
> 
> > > > > > > > > 
> 
> > > > > > > > >       
> 
> > > > > > > > >     
> 
> > > > > > > > > 
> 
> > > > > > > > >     
> 
> > > > > > > > >   
> 
> > > > > > > > > The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.
> 
> > > > > > > > 
> 
> > > > > > > > 
> 
> > > > > > > > This shouldn't be in my proof. I apologize: it's actually malformed IR.
> 
> > > > > > > > nuw is impossible for x ugt -C ? x + C : -1;
> 
> > > > > > > 
> 
> > > > > > > 
> 
> > > > > > > Why would it be malformed IR? The select would pick -1 in the case of overflow and the add would produce poison. The selects blocks the poison from propagating by picking -1. That's completely valid IR
> 
> > > > > > 
> 
> > > > > > 
> 
> > > > > > LLVM would NEVER put nuw for x + C if x ugt -C was true.
> 
> > > > > > Let C1 be the positive interpretation of -C in two's compliment.
> 
> > > > > > x ugt C1 being true means x is at (unsigned least) C1 + 1. But, C + C1 would wrap. meaning X + C MUST wrap if X ugt -C.
> 
> > > > > 
> 
> > > > > 
> 
> > > > > Ignoring the undef input possibility, alive2 says its valid to apply nuw. https://alive2.llvm.org/ce/z/sPStwP
> 
> > > > 
> 
> > > > 
> 
> > > > Okay thank you. Applying fixes right now.
> 
> > > 
> 
> > > 
> 
> > > What fix are you applying?
> 
> > 
> 
> > I am going to apply the nuw stuff. I just need to update my tests
> 
> 
> 
> Done!

I was not asking for that change. I was only arguing about whether the IR was malformed.

RSilicon commented:
> > > > This PR was pretty close to an acceptable state the last time I looked at it. Now you have once again added extra changes that nobody asked for, and moved it away from acceptance again
> 
> > 
> 
> > > 
> 
> > 
> 
> > > > > > > > > > Your alive proof has
> 
> > 
> 
> > > > > > > > > > ```
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > > define dso_local i8 @src3(i8 %x, i8  %c) {
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > > entry:
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > >   %noZero = icmp ne i8 %c, 0
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > >   call void @llvm.assume(i1 %noZero)
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > >   %add = add nuw i8 %x, %c
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > >   %c.not = xor i8 %c, -1
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > >   %d = add i8 %c.not, 1
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > >   %cmp.not = icmp ugt i8 %x, %d
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > >   %cond = select i1 %cmp.not, i8 -1, i8 %add
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > >   ret i8 %cond
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > > }
> 
> > 
> 
> > > > > > > > > > ```
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > >     
> 
> > 
> 
> > > > > > > > > >       
> 
> > 
> 
> > > > > > > > > >     
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > >       
> 
> > 
> 
> > > > > > > > > >     
> 
> > 
> 
> > > > > > > > > > 
> 
> > 
> 
> > > > > > > > > >     
> 
> > 
> 
> > > > > > > > > >   
> 
> > 
> 
> > > > > > > > > > The `nuw` flag there says that the add never overflows and thus never saturates. The transform fails without the nuw. There's no check for the nuw in your transform.
> 
> > 
> 
> > > > > > > > > 
> 
> > 
> 
> > > > > > > > > 
> 
> > 
> 
> > > > > > > > > This shouldn't be in my proof. I apologize: it's actually malformed IR.
> 
> > 
> 
> > > > > > > > > nuw is impossible for x ugt -C ? x + C : -1;
> 
> > 
> 
> > > > > > > > 
> 
> > 
> 
> > > > > > > > 
> 
> > 
> 
> > > > > > > > Why would it be malformed IR? The select would pick -1 in the case of overflow and the add would produce poison. The selects blocks the poison from propagating by picking -1. That's completely valid IR
> 
> > 
> 
> > > > > > > 
> 
> > 
> 
> > > > > > > 
> 
> > 
> 
> > > > > > > LLVM would NEVER put nuw for x + C if x ugt -C was true.
> 
> > 
> 
> > > > > > > Let C1 be the positive interpretation of -C in two's compliment.
> 
> > 
> 
> > > > > > > x ugt C1 being true means x is at (unsigned least) C1 + 1. But, C + C1 would wrap. meaning X + C MUST wrap if X ugt -C.
> 
> > 
> 
> > > > > > 
> 
> > 
> 
> > > > > > 
> 
> > 
> 
> > > > > > Ignoring the undef input possibility, alive2 says its valid to apply nuw. https://alive2.llvm.org/ce/z/sPStwP
> 
> > 
> 
> > > > > 
> 
> > 
> 
> > > > > 
> 
> > 
> 
> > > > > Okay thank you. Applying fixes right now.
> 
> > 
> 
> > > > 
> 
> > 
> 
> > > > 
> 
> > 
> 
> > > > What fix are you applying?
> 
> > 
> 
> > > 
> 
> > 
> 
> > > I am going to apply the nuw stuff. I just need to update my tests
> 
> > 
> 
> > 
> 
> > 
> 
> > Done!
> 
> 
> 
> I was not asking for that change. I was only arguing about whether the IR was malformed. I think I was off by 1 in my alive proof anyway. Please remove the NUW change.

We were both wrong. I was correct but it turns out there were two other cases I missed that did involve NUW, and I included those new cases, and wrote tests for them. Take a look for yourself.

AZero13 committed:
[InstCombine] Canonicalize more saturated-add variants

LLVM is not evaluating X u > C, a, b the same way it evaluates X <= C, b, a.

To fix this, let's move the folds to after the canonicalization of -1 to TrueVal.

Let's allow splat vectors with poison elements to be recognized too!

Finally, for completion, handle the one case that isn't caught by the above checks because it is canonicalized to eq:
X == -1 ? -1 : X + 1 -> uadd.sat(X, 1)

Alive2 Proof:
https://alive2.llvm.org/ce/z/WEcgYH

RSilicon commented:
@topperc Do you have any objections to this PR as it is of now?

topperc commented:
> @topperc Do you have any objections to this PR as it is of now?

Remove the nuw handling unless you can show a real world test that benefits from it. We don't have to solve every possible case in InstCombine.

RSilicon commented:
> > @topperc Do you have any objections to this PR as it is of now?
> 
> Remove the nuw handling unless you can show a real world test that benefits from it. We don't have to solve every possible case in InstCombine.

X > -C ? -1 : X + C is the canonicalization of the literal definition of sat.add.

X + C has to be nuw for this to work. 

X >= -C + 1 is basically a weird version of this.

topperc commented:
> > > @topperc Do you have any objections to this PR as it is of now?
> 
> > 
> 
> > Remove the nuw handling unless you can show a real world test that benefits from it. We don't have to solve every possible case in InstCombine.
> 
> 
> 
> X > -C ? -1 : X + C is the canonicalization of the literal definition of sat.add.
> 
> 
> 
> X + C has to be nuw for this to work. 
> 
> 
> 
> X >= -C + 1 is basically a weird version of this. 

Write a C function that generates IR with nuw on the add.

RSilicon commented:
> > > > @topperc Do you have any objections to this PR as it is of now?
> > 
> > 
> > > 
> > 
> > 
> > > Remove the nuw handling unless you can show a real world test that benefits from it. We don't have to solve every possible case in InstCombine.
> > 
> > 
> > X > -C ? -1 : X + C is the canonicalization of the literal definition of sat.add.
> > X + C has to be nuw for this to work.
> > X >= -C + 1 is basically a weird version of this.
> 
> Write a C function that generates IR with nuw on the add.
<img width="1371" alt="Screenshot 2024-07-30 at 2 39 05 PM" src="https://github.com/user-attachments/assets/fce8b5c4-4b3b-4d75-9275-062a0d79ba12">

RSilicon commented:
> > > > @topperc Do you have any objections to this PR as it is of now?
> > 
> > 
> > > 
> > 
> > 
> > > Remove the nuw handling unless you can show a real world test that benefits from it. We don't have to solve every possible case in InstCombine.
> > 
> > 
> > X > -C ? -1 : X + C is the canonicalization of the literal definition of sat.add.
> > X + C has to be nuw for this to work.
> > X >= -C + 1 is basically a weird version of this.
> 
> Write a C function that generates IR with nuw on the add.

Okay you were right. I am removing the NUW add case.

RSilicon commented:
Ugh I merged the removal of the unneeded tests into the wrong commit.

This shouldn't be a problem though, right? @topperc

RSilicon commented:
Do any of you have any objections to this PR as it is of now?

topperc commented on llvm/test/Transforms/InstCombine/saturating-add-sub.ll:
```diff
@@ -1395,6 +1395,348 @@ define i32 @uadd_sat(i32 %x, i32 %y) {
+  ret i32 %cond
+}
+
+define i32 @uadd_sat_negative_one_poison(i32 %x) {
```
I don't know if the scalar poison tests are interesting. The icmp with poison operand will folded to poison before the select is visited. The add with poison operand will also be folded to poison before the select is visited.

RSilicon replied:
> I don't know if the scalar poison tests are interesting. The icmp with poison operand will folded to poison before the select is visited. The add with poison operand will also be folded to poison before the select is visited.

Addressed!

nikic replied:
It doesn't look like this was done. There are still 9 tests with scalar poison.

RSilicon commented:
Do any of you have any objections to this PR as it is of now?

RSilicon commented:
@topperc I believe this is ready now!

RSilicon commented:
@topperc Is this PR ready to merge?

topperc commented:
> @topperc Is this PR ready to merge?

Please do not ping a patch more than once per week per the coding standards. https://llvm.org/docs/CodeReview.html#lgtm-how-a-patch-is-accepted

```
Note that, if a reviewer has requested a particular community member to review, and after a week that community member has yet to respond, feel free to ping the patch (which literally means submitting a comment on the patch with the word, “Ping.”), or alternatively, ask the original reviewer for further suggestions.
```

topperc reviewed (approved):
LGTM

RSilicon commented:
> > @topperc Is this PR ready to merge?
> 
> 
> 
> Please do not ping a patch more than once per week per the coding standards. https://llvm.org/docs/CodeReview.html#lgtm-how-a-patch-is-accepted
> 
> 
> 
> "Note that, if a reviewer has requested a particular community member to review, and after a week that community member has yet to respond, feel free to ping the patch (which literally means submitting a comment on the patch with the word, “Ping.”), or alternatively, ask the original reviewer for further suggestions."

Understood!

topperc commented on llvm/test/Transforms/InstCombine/saturating-add-sub.ll:
```diff
@@ -1395,6 +1395,338 @@ define i32 @uadd_sat(i32 %x, i32 %y) {
+; CHECK-NEXT:    ret i32 -1
+;
+  %cmp = icmp eq i32 %x, -1
+  %add = add i32 %x, poison
```
Still testing scalar poison

RSilicon replied:
Addressed!

RSilicon replied:
> Still testing scalar poison

Do you think I should replace them or just remove them?

RSilicon replied:
> Still testing scalar poison

Addressed!

topperc reviewed (approved):
LGTM

RSilicon commented:
> LGTM

Can we please merge this?

RSilicon commented:
@dtcxzyw Could you please merge this as I do not have commit perms?

topperc commented:
Why do you force push patches after they have been approved? We have to re-review to make sure nothing changed. But force push makes that hard.

RSilicon commented:
> Why do you force push patches after they have been approved? We have to re-review to make sure nothing changed. But force push makes that hard.

I folded the third commit away.

topperc commented:
Committed as da140bc1e3be4dfd49e17c2f4cc496ae75ff8693 and ede49fa5f3f32796618a71880038e2564c9d4db9
